code/datums/components/surgery_initiator.dm

```diff
@@ -0,0 +1,159 @@
+/**
+  *
+  * Allows parent (obj) to initiate surgeries.
+  *
+  */
+/datum/component/surgery_initiator
+	dupe_mode = COMPONENT_DUPE_UNIQUE
+	///allows for post-selection manipulation of parent
+	var/datum/callback/after_select_cb
+
+/datum/component/surgery_initiator/Initialize(_after_select_cb)
+	. = ..()
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+	after_select_cb = _after_select_cb
+
+/datum/component/surgery_initiator/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ITEM_ATTACK, .proc/initiate_surgery_moment)
+
+/datum/component/surgery_initiator/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_ITEM_ATTACK)
+
+/datum/component/surgery_initiator/Destroy()
+	if(after_select_cb)
+		QDEL_NULL(after_select_cb)
+	return ..()
+
+	/**
+	  *
+	  * Does the surgery initiation.
+	  *
+	  */
+/datum/component/surgery_initiator/proc/initiate_surgery_moment(datum/source, atom/target, mob/user)
+	if(!isliving(target))
+		return
+	var/mob/living/livingtarget = target
+	var/mob/living/carbon/carbontarget
+	var/obj/item/bodypart/affecting
+	var/selected_zone = user.zone_selected
+	. = COMPONENT_ITEM_NO_ATTACK
+
+	if(iscarbon(livingtarget))
+		carbontarget = livingtarget
+		affecting = carbontarget.get_bodypart(check_zone(selected_zone))
+
+	var/datum/surgery/current_surgery
+
+	for(var/i_one in livingtarget.surgeries)
+		var/datum/surgery/surgeryloop = i_one
+		if(surgeryloop.location == selected_zone)
+			current_surgery = surgeryloop
+
+	if(!current_surgery)
+		var/list/all_surgeries = GLOB.surgeries_list.Copy()
+		var/list/available_surgeries = list()
+
+		for(var/i_two in all_surgeries)
+			var/datum/surgery/surgeryloop_two = i_two
+			if(!surgeryloop_two.possible_locs.Find(selected_zone))
+				continue
+			if(affecting)
+				if(!surgeryloop_two.requires_bodypart)
+					continue
+				if(surgeryloop_two.requires_bodypart_type && affecting.status != surgeryloop_two.requires_bodypart_type)
+					continue
+				if(surgeryloop_two.requires_real_bodypart && affecting.is_pseudopart)
+					continue
+			else if(carbontarget && surgeryloop_two.requires_bodypart) //mob with no limb in surgery zone when we need a limb
+				continue
+			if(surgeryloop_two.lying_required && (livingtarget.mobility_flags & MOBILITY_STAND))
+				continue
+			if(!surgeryloop_two.can_start(user, livingtarget))
+				continue
+			for(var/path in surgeryloop_two.target_mobtypes)
+				if(istype(livingtarget, path))
+					available_surgeries[surgeryloop_two.name] = surgeryloop_two
+					break
+
+		if(!available_surgeries.len)
+			return
+
+		var/pick_your_surgery = input("Begin which procedure?", "Surgery", null, null) as null|anything in sortList(available_surgeries)
```

Comment:
you cannot sleep in signal listeners, they are called with waitfor set to 0

Reply:
this causes a race condition for the return value of the signal.

Reply:
suggest you use getComponent and initiate a proc call directly on the component.

Reply:
You can also invoke an async proc that sleeps for your timed action and guess a return value so things go on.
We should make signals async, really.

Reply:
if you are both 
* sleeping in the proc
* and a signal

to me that's a sign you are trying to do too much over a signal boundary, and need to be using getcomponent instead.

Signals should not end up as a replacement for fucking proc calls

Reply:
oh and also, that effectively makes every single send_signal a sleep, which means you need to recheck all your conditions after it.

Big bad imo

Reply:
Well, potential one, yea.
Dunno, I see your point. There's pros and cons for either side.

Reply:
so if I wanted to make potentially any obj/item a surgery initiator, I need to directly check for the component on obj/item attack/afterattack?

Reply:
Or not use components for such a thing.

Reply:
Do what I said before: once the `do_after()` begins invoke that as a separate async proc and return a value ending the attack chain